### PR TITLE
[CIS-259] Implement teams

### PR DESCRIPTION
**Pull request by DominikBucher12**
# What this PR do:
- Adds support for user teams and channel team inside those payloads
# How to test it
- Run tests if `teams` property is parsed from payload.
# Where you can start
- `UserPayload` and `ChannelListPayload`, check if there is correct relationship between objects.

### CODE REVIEW

**Review by VojtaStavik — CHANGES_REQUESTED** [review]
Ideally, the tests should be in the same commit as the implementation, so it's easy to verify what's happening there.

It seems there are no tests for DTOs, correct? I see only tests for the payload serialization.

File: Sources/StreamChat/Database/DTOs/UserDTO.swift [hunk 1]
```diff
@@ -108,8 +109,9 @@ extension NSManagedObjectContext: UserDatabaseSession {
         dto.userCreatedAt = payload.createdAt
         dto.userRoleRaw = payload.role.rawValue
         dto.userUpdatedAt = payload.updatedAt
-
-        // TODO: TEAMS
+        dto.teams = try payload.teams
+            .map { try saveTeam(teamId: $0) }
+            .reduce(into: Set<TeamDTO>()) { teamSet, team in teamSet.insert(team) }
```
[the remark this answers is not in the document]
Comment (LukasHromadnik) [reply Sources/StreamChat/Database/DTOs/UserDTO.swift:114]: IMHO for the best readability I would do
```
let teams = try payload.teams.map { try saveTeam(teamId: $0) }
dto.teams = Set(teams)
```
you can also do it on one line without the `teams` variable

File: Sources/StreamChat/Models/User.swift [hunk 2]
```diff
@@ -81,7 +84,7 @@ public class _ChatUser<ExtraData: UserExtraData> {
         createdAt: Date = .init(),
         updatedAt: Date = .init(),
         lastActiveAt: Date? = nil,
-        teams: [String] = [],
+        teams: [TeamId] = [],
         extraData: ExtraData = .defaultValue
```
Comment (nuno-vieira) [inline Sources/StreamChat/Models/User.swift:88]: If TeamDTO includes users and channels, does it make sense for us to only include the TeamId on our domain models?

File: Tests/StreamChatTests/MockEnpointResponses/CurrentUser.json [hunk 3]
```diff
@@ -12,6 +12,11 @@
   "online" : true,
   "name" : "Broken Waterfall",
   "secret_note" : "Anaking is Vader!",
+  "teams" : [
+      "RED",
+      "GREEN",
+      "BLUE"
+  ],
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/MockEnpointResponses/CurrentUser.json:19]: According to the TeamDTO, it can also include users and channels? Or is that only in the DTO?

**Review by nuno-vieira — APPROVED** [review]
Great Job 💪 Just added some questions 👍

File: Tests/StreamChatTests/MockEnpointResponses/CurrentUser.json (hunk 3 above)
Comment (DominikBucher12) [reply Tests/StreamChatTests/MockEnpointResponses/CurrentUser.json:19]: Yes, however those properties are not fields of the payload gotten from response. We should handle the connections inside our codebase aka connect the team with the user (or channel) when assigning team/teams to those.

File: Sources/StreamChat/Models/User.swift (hunk 2 above)
Comment (DominikBucher12) [reply Sources/StreamChat/Models/User.swift:88]: -> Internal discussion on Slack.

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        // Asynchronously save the user payload to the db, which should save teams for user as well.
+        database.write { session in
+            try! session.saveUser(payload: payload)
+        }
+
+        // Asynchronously save the payload to the db
+        database.write { session in
+            try! session.saveUser(payload: payload)
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:35]: Why is this saved twice?

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        // Load the user from the db and check the fields are correct
+        var loadedUserDTO: UserDTO? {
+            database.viewContext.user(id: userId)
+        }
+
+        AssertAsync {
+            Assert.willBeEqual(payload.teams.count, loadedUserDTO?.teams?.count)
+            // It's possible that `loadedUserDTO` is nil so it would sabotage next test.
+            // Let's check this first so we can check channel relationship on next line
+            Assert.willNotBeNil(loadedUserDTO?.teams?.first?.users.first)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:46]: That's exactly why we have `database.writeSynchronously` ;-) Using that you can convert this to normal `XCTAssert`s

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            Assert.willBeEqual(loadedUserDTO?.teams?.first?.users.first, loadedUserDTO)
+        }
+    }
+
+    func test_teamForChannel_isStoredAndLoadedFromDB() {
+        let teamId = UUID().uuidString
+        let channelId = try! ChannelId(cid: "messaging:team")
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:53]: Why not just
```swift
let teamId: TeamId = .unique
let channelId: ChannelId = .uniqe
```
?

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        let channelDetailPayload: ChannelDetailPayload<NoExtraData> = .dummy(cid: channelId, teamId: teamId)
+
+        let channelPayload: ChannelPayload<NoExtraData> = .init(
+            channel: channelDetailPayload, watcherCount: 0,
+            watchers: nil,
+            members: [],
+            membership: nil,
+            messages: [],
+            channelReads: []
+        )
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:64]: Do you need to create this wrapper payload? IMO you can use `session.saveChannel` with `ChannelDetailPayload`, too

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            database.viewContext.channel(cid: channelId)
+        }
+
+        AssertAsync {
+            Assert.willBeEqual(channelPayload.channel.team, loadedChannelDTO?.team?.id)
+            // It's possible that `loadedChannelDTO` is nil so it would sabotage next test.
+            // Let's check this first so we can check channel relationship on next line
+            Assert.willNotBeNil(loadedChannelDTO?.team?.channels.first)
+            Assert.willBeEqual(loadedChannelDTO?.team?.channels.first, loadedChannelDTO)
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:82]: Just nit: The tests inside `AssertAsync { }` are evaluated in parallel. To achieve what you wanted to achieve, you'd have to do:
```
AssertAsync.willNotBeNil(loadedChannelDTO?.team?.channels.first)
AssertAsync.willBeEqual(loadedChannelDTO?.team?.channels.first, loadedChannelDTO)
```

File: Sources/StreamChat/Database/DTOs/UserDTO.swift [hunk 9]
```diff
@@ -109,15 +110,19 @@ extension NSManagedObjectContext: UserDatabaseSession {
...
         dto.userUpdatedAt = payload.updatedAt
 
-        // TODO: TEAMS
-
         dto.extraData = try JSONEncoder.default.encode(payload.extraData)
-        
+
+        let teamsArray: [TeamDTO] = try payload.teams.map {
+            let team = try saveTeam(teamId: $0)
+            team.users.insert(dto)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO.swift:117]: this is IMO not needed. The relationship is already set on L120

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 10]
```diff
@@ -126,13 +126,21 @@ extension NSManagedObjectContext {
         dto.isFrozen = payload.isFrozen
         dto.cooldownDuration = payload.cooldownDuration
 
+        if let teamId = payload.team {
+            let teamDTO = try saveTeam(teamId: teamId)
+            teamDTO.channels.insert(dto)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:131]: again, not needed

File: Tests/Shared/CustomAssertions/AssertAsync.swift [hunk 11]
```diff
@@ -93,6 +93,32 @@ extension Assert {
...
+        line: UInt = #line
+    ) -> Assertion {
+        willBeTrue(
+            expression1() != nil,
+            timeout: timeout,
+            message: "Failed to not be `nil`",
+            file: file,
+            line: line
+        )
+    }
```
Comment (VojtaStavik) [inline Tests/Shared/CustomAssertions/AssertAsync.swift:121]: Ideally this should be a separate commit

File: CHANGELOG.md [hunk 12]
```diff
@@ -19,6 +19,7 @@ _March 12, 2021_
 - Expose `membership` value on `ChatChannel` which contains information about the current user membership [#885](https://github.com/GetStream/stream-chat-swift/issues/885)
 - `ChatChannelMember` now contains channel-specific ban information: `isBannedFromChannel` and `banExpiresAt` [#885](https://github.com/GetStream/stream-chat-swift/issues/885)
 - Channel-specific ban events are handled and the models are properly updated [#885](https://github.com/GetStream/stream-chat-swift/pull/885)
+- Introduce support for [multitenancy](https://getstream.io/chat/docs/react/multi_tenant_chat/?language=swift) - `teams` for `User` and `team` for `Channel` are now parsed from payloads.
```
Comment (VojtaStavik) [inline CHANGELOG.md:22]: `are now parsed from payloads.` -> `are now exposed`? ;-)

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 13]
```diff
@@ -126,13 +126,21 @@ extension NSManagedObjectContext {
...
+            dto.team = teamDTO
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:132]: Is the `ChatChannel.team` value tested somewhere?

**Review by VojtaStavik — CHANGES_REQUESTED** [review]
Almost there... just some nits

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift (hunk 13 above)
Comment (DominikBucher12) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:131]: oh my zshell you are right! I forgot to uncomment the property on the model. Thanks 🍻

File: CHANGELOG.md (hunk 12 above)
Comment (DominikBucher12) [reply CHANGELOG.md]: Yes.

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+        // Load the user from the db and check the fields are correct
+        var loadedUserDTO: UserDTO? {
+            database.viewContext.user(id: userId)
+        }
+
+        XCTAssertEqual(payload.teams.count, loadedUserDTO?.teams?.count)
+        XCTAssertEqual(loadedUserDTO?.teams?.first?.users.first, loadedUserDTO)
+    }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:39]: This should be among `UserDTO` tests

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        // Load the channel from the db and check the team is correct
+        var loadedChannelDTO: ChannelDTO? {
+            database.viewContext.channel(cid: channelId)
+        }
+
+        XCTAssertEqual(channelDetailPayload.team, loadedChannelDTO?.team?.id)
+        XCTAssertEqual(loadedChannelDTO?.team?.channels.first, loadedChannelDTO)
+    }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:59]: This should be among `ChannelDTO` tests

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    func test_teamsForUser_areStoredAndLoadedFromDB() {
+        let teamIds: [TeamId] = [.unique, .unique, .unique]
+        let userId: UserId = .unique
+
+        let payload: UserPayload<NoExtraData> = .dummy(userId: userId, teams: teamIds)
+
+        // Asynchronously save the user payload to the db, which should save teams for user as well.
+        try! database.writeSynchronously { session in
+            try! session.saveUser(payload: payload)
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:30]: If you mark the test `throws` you can cache `try!` to `try`

File: Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+        // Load the user from the db and check the fields are correct
+        var loadedUserDTO: UserDTO? {
+            database.viewContext.user(id: userId)
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/TeamDTO_Tests.swift:35]:
```
let loadedUser = try XCTUnwrap(database.viewContext.user(id: userId))
```
would work better here

File: Sources/StreamChat/Database/DTOs/UserDTO.swift (hunk 9 above)
Comment (VojtaStavik) [reply Sources/StreamChat/Database/DTOs/UserDTO.swift:117]: So will you remove it then? :trollface:

File: Tests/StreamChatTests/DummyData/ChannelPayload.swift [hunk 18]
```diff
@@ -7,7 +7,7 @@ import Foundation
 
 extension ChannelDetailPayload where ExtraData == NoExtraData {
     /// Returns a dummy channel detail payload with the given cid
-    static func dummy(cid: ChannelId) -> ChannelDetailPayload {
+    static func dummy(cid: ChannelId, teamId: TeamId? = nil) -> ChannelDetailPayload {
```
Comment (VojtaStavik) [inline Tests/StreamChatTests/DummyData/ChannelPayload.swift:10]: Not sure if this is needed tbh :) Why no just assign `.unique` to the `team` variable like we do it with other things?

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 19]
```diff
@@ -43,6 +44,8 @@ class UserDTO_Tests: XCTestCase {
             Assert.willBeEqual(payload.createdAt, loadedUserDTO?.userCreatedAt)
             Assert.willBeEqual(payload.updatedAt, loadedUserDTO?.userUpdatedAt)
             Assert.willBeEqual(payload.lastActiveAt, loadedUserDTO?.lastActivityAt)
+            Assert.willBeEqual(payload.teams.first, loadedUserDTO?.teams?.first?.id)
+            Assert.willBeEqual(loadedUserDTO?.teams?.first?.users.first, loadedUserDTO)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:48]: What exactly are you testing here? 🤯 This is already taken care of by CoreData, we don't need to test it explicitly.

File: Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift [hunk 20]
```diff
@@ -43,6 +44,8 @@ class UserDTO_Tests: XCTestCase {
             Assert.willBeEqual(payload.createdAt, loadedUserDTO?.userCreatedAt)
             Assert.willBeEqual(payload.updatedAt, loadedUserDTO?.userUpdatedAt)
             Assert.willBeEqual(payload.lastActiveAt, loadedUserDTO?.lastActivityAt)
+            Assert.willBeEqual(payload.teams.first, loadedUserDTO?.teams?.first?.id)
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/UserDTO_Tests.swift:47]: We can be more specific here:
```
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -20,6 +20,7 @@ _March 12, 2021_
 - Expose `membership` value on `ChatChannel` which contains information about the current user membership [#885](https://github.com/GetStream/stream-chat-swift/issues/885)
 - `ChatChannelMember` now contains channel-specific ban information: `isBannedFromChannel` and `banExpiresAt` [#885](https://github.com/GetStream/stream-chat-swift/issues/885)
 - Channel-specific ban events are handled and the models are properly updated [#885](https://github.com/GetStream/stream-chat-swift/pull/885)
+- Introduce support for [multitenancy](https://getstream.io/chat/docs/react/multi_tenant_chat/?language=swift) - `teams` for `User` and `team` for `Channel` are now exposed.
```
Comment (VojtaStavik) [inline CHANGELOG.md:23]: The link to the PR is missing (see other CHANGELOG items please)

**Review by VojtaStavik — APPROVED** [review]
Cool 🎉 There's are still some strange things happening in tests. Please fix and it's good to 🚢
